tempesta_fw/cache.c

```diff
@@ -1847,54 +1886,138 @@ tfw_cache_purge_method(TfwHttpReq *req)
+		off = 0;
+		frame_hdr->flags = last_frag ? HTTP2_F_END_STREAM : 0;
+		if (*first_frag) {
+			frame_hdr->length = sz - FRAME_HEADER_SIZE;
+			memcpy_fast(new_p, p, sz);
+		}
+		else {
+			frame_hdr->length = sz;
+			memcpy_fast(new_p + FRAME_HEADER_SIZE, p, sz);
+			sz += FRAME_HEADER_SIZE;
+		}
+		tfw_h2_pack_frame_header(new_p, frame_hdr);
+	}
+	else {
+		off = ((unsigned long)p & ~PAGE_MASK);
+		if (*first_frag) {
+			off += FRAME_HEADER_SIZE;
+			sz -= FRAME_HEADER_SIZE;
+		}
+		page = virt_to_page(p);
+	}
+
+	skb_fill_page_desc(it->skb, it->frag, page, off, sz);
+	skb_frag_ref(it->skb, it->frag);
+	ss_skb_adjust_data_len(it->skb, sz);
+	++it->frag;
+
+	*first_frag = false;
+
+	return 0;
+}
+
 /**
  * Build the message body as paged fragments of skb.
  * See do_tcp_sendpages() as reference.
  */
 static int
-tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p)
+tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p,
+			  unsigned long body_sz, bool h2, unsigned int stream_id)
 {
 	int r;
+	TfwFrameHdr frame_hdr = {.stream_id = stream_id, .type = HTTP2_DATA};
+	bool first_frag = true;
 
 	if (WARN_ON_ONCE(!it->skb_head))
 		return -EINVAL;
 	/*
 	 * If all skbs/frags are used up (see @tfw_http_msg_expand_data()),
 	 * create new skb with empty frags to reference the cached body;
-	 * otherwise, use next empty frag in current skb.
+	 * otherwise, use next empty frag in current skb. Constructing h2
+	 * responses from cache always imply data copies, for SKBTX_SHARED_FRAG
+	 * skbs, copy will happen twice. If body is relatively small keep
+	 * using existing skb and allow double copy, otherwise allocate a new
+	 * skb even if previous is not fully populated.
 	 */
-	if (!it->skb) {
+	if (!it->skb || (it->frag >= MAX_SKB_FRAGS)
+	    || (h2 && (skb_shinfo(it->skb)->tx_flags & SKBTX_SHARED_FRAG)
+		&& body_sz > PAGE_SIZE))
+	{
 		if  ((r = tfw_msg_iter_append_skb(it)))
 			return r;
-	} else {
-		++it->frag;
-		if (it->frag >= MAX_SKB_FRAGS
-		    && (r = tfw_msg_iter_append_skb(it)))
-			return r;
+		if (h2)
+			skb_shinfo(it->skb)->tx_flags &= ~SKBTX_SHARED_FRAG;
 	}
-	BUG_ON(it->frag < 0);
+	if (WARN_ON_ONCE(it->frag < 0))
+		return -EINVAL;
 
 	while (1) {
-		int off, f_size;
+		int off, f_size, min_sz = first_frag ? FRAME_HEADER_SIZE : 0;
 
-		/* TDB keeps data by pages and we can reuse the pages. */
 		off = (unsigned long)p & ~PAGE_MASK;
 		f_size = trec->data + trec->len - p;
-		if (f_size) {
-			skb_fill_page_desc(it->skb, it->frag, virt_to_page(p),
-					   off, f_size);
-			skb_frag_ref(it->skb, it->frag);
-			ss_skb_adjust_data_len(it->skb, f_size);
-			++it->frag;
+		/*
+		 * The body is stored in native h2 format, with reserved space
+		 * for h2 headers. Skip page fragment if it's too short to store
+		 * header.
+		 */
+		if (f_size > min_sz) {
+			body_sz -= f_size;
+			r = tfw_cache_add_body_page(it, p, f_size, &frame_hdr,
+						    h2, &first_frag, !body_sz);
```

Comment:
The `first_frag` variable is changed only once in the cycle: inside `tfw_cache_add_body_page()` and in any case. Thus, it seems  that we can pass the `first_frag` into the procedure just by value (instead of passing by pointer) and change it to `false` here, in the caller procedure - right after the successful `tfw_cache_add_body_page()` call.

Reply:
The whole `first_frag` was removed, so mark this comment as fixed.